Nomad panic when starting in dev mode with consul and vault on same host

### Nomad version
```
root@precision-5520:~# nomad version
Nomad v0.7.0-beta1
```

### Operating system and Environment details
```
root@precision-5520:~# lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux unstable (sid)
Release:	unstable
Codename:	sid
root@precision-5520:~# uname -a
Linux precision-5520 4.13.0-1-amd64 #1 SMP Debian 4.13.4-1 (2017-10-01) x86_64 GNU/Linux
```

### Issue
```
root@precision-5520:~# nomad agent -dev
    No configuration files loaded
==> Starting Nomad agent...
panic: interface conversion: interface {} is nil, not bool

goroutine 1 [running]:
github.com/hashicorp/nomad/client/fingerprint.(*ConsulFingerprint).Fingerprint(0xc4202e8680, 0xc420090000, 0xc4204d8000, 0x1eeac20, 0xc4202e8680, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/client/fingerprint/consul.go:68 +0xddd
github.com/hashicorp/nomad/client.(*Client).fingerprint(0xc420090160, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/client/client.go:899 +0x56d
github.com/hashicorp/nomad/client.NewClient(0xc420090000, 0x1eeb560, 0xc42000e040, 0x1ef36e0, 0xc4205520c0, 0xc4201bd400, 0xc4201e0018, 0xc420436180, 0xc4201e0010)
	/opt/gopath/src/github.com/hashicorp/nomad/client/client.go:229 +0x7a9
github.com/hashicorp/nomad/command/agent.(*Agent).setupClient(0xc4201ef030, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/command/agent/agent.go:503 +0x14a
github.com/hashicorp/nomad/command/agent.NewAgent(0xc4201feb40, 0x1ee6860, 0xc42044e8e0, 0xc4201bd360, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/command/agent/agent.go:90 +0x227
github.com/hashicorp/nomad/command/agent.(*Command).setupAgent(0xc4201da380, 0xc4201feb40, 0x1ee6860, 0xc42044e8e0, 0xc4201bd360, 0xc42044e8e0, 0xc42001d300)
	/opt/gopath/src/github.com/hashicorp/nomad/command/agent/command.go:337 +0x96
github.com/hashicorp/nomad/command/agent.(*Command).Run(0xc4201da380, 0xc4200100b0, 0x1, 0x1, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/command/agent/command.go:455 +0x3ea
github.com/hashicorp/nomad/vendor/github.com/mitchellh/cli.(*CLI).Run(0xc4201fe140, 0xc4201fe140, 0x3d, 0x1573178)
	/opt/gopath/src/github.com/hashicorp/nomad/vendor/github.com/mitchellh/cli/cli.go:255 +0x1eb
main.RunCustom(0xc4200100a0, 0x2, 0x2, 0xc4201d9890, 0xc4204d9f60)
	/opt/gopath/src/github.com/hashicorp/nomad/main.go:60 +0x1245
main.Run(0xc4200100a0, 0x2, 0x2, 0x1182e26)
	/opt/gopath/src/github.com/hashicorp/nomad/main.go:24 +0x56
main.main()
	/opt/gopath/src/github.com/hashicorp/nomad/main.go:20 +0x63
```

### Reproduction steps
I am going on a blitz here and running a consul server agent in dev mode, a vault server in dev mode and attempting to get nomad running in dev mode on my workstation.  

I downloaded the nomad 0.6.3 binary as well and got the same result.

```
matthew@precision-5520:~/Workspace/vault-tutorial$ vault version
Vault v0.8.3 ('6b29fb2b7f70ed538ee2b3c057335d706b6d4e36')
matthew@precision-5520:~/Workspace/vault-tutorial$ consul version
Consul v1.0.0-beta1-beta1 (82f312d)
Protocol 2 spoken by default, understands 2 to 3 (agent will automatically use protocol >2 when speaking to compatible agents)
matthew@precision-5520:~/Workspace/vault-tutorial$ nomad version
Nomad v0.7.0-beta1
```

### Nomad Server logs (if appropriate)
Couldn't find any of note.

### Nomad Client logs (if appropriate)
Couldn't find any of note.

### Job file (if appropriate)
N/A

Any other questions or items needed, let me know!
Matthew

--- a/client/fingerprint/consul.go
+++ b/client/fingerprint/consul.go
@@ -65,15 +65,39 @@ func (f *ConsulFingerprint) Fingerprint(config *client.Config, node *structs.Nod
 		return false, nil
 	}
 
-	node.Attributes["consul.server"] = strconv.FormatBool(info["Config"]["Server"].(bool))
-	node.Attributes["consul.version"] = info["Config"]["Version"].(string)
-	node.Attributes["consul.revision"] = info["Config"]["Revision"].(string)
-	node.Attributes["unique.consul.name"] = info["Config"]["NodeName"].(string)
-	node.Attributes["consul.datacenter"] = info["Config"]["Datacenter"].(string)
+	if s, ok := info["Config"]["Server"].(bool); ok {
+		node.Attributes["consul.server"] = strconv.FormatBool(s)
+	} else {
+		f.logger.Printf("[WARN] fingerprint.consul: unable to fingerprint consul.server")
+	}
+	if v, ok := info["Config"]["Version"].(string); ok {
+		node.Attributes["consul.version"] = v
+	} else {
+		f.logger.Printf("[WARN] fingerprint.consul: unable to fingerprint consul.version")
+	}
+	if r, ok := info["Config"]["Revision"].(string); ok {
+		node.Attributes["consul.revision"] = r
+	} else {
+		f.logger.Printf("[WARN] fingerprint.consul: unable to fingerprint consul.revision")
+	}
+	if n, ok := info["Config"]["NodeName"].(string); ok {
+		node.Attributes["unique.consul.name"] = n
+	} else {
+		f.logger.Printf("[WARN] fingerprint.consul: unable to fingerprint unique.consul.name")
+	}
+	if d, ok := info["Config"]["Datacenter"].(string); ok {
+		node.Attributes["consul.datacenter"] = d
+	} else {
+		f.logger.Printf("[WARN] fingerprint.consul: unable to fingerprint consul.datacenter")
+	}
 
-	node.Links["consul"] = fmt.Sprintf("%s.%s",
-		node.Attributes["consul.datacenter"],
-		node.Attributes["unique.consul.name"])
+	if node.Attributes["consul.datacenter"] != "" || node.Attributes["unique.consul.name"] != "" {
+		node.Links["consul"] = fmt.Sprintf("%s.%s",
+			node.Attributes["consul.datacenter"],
+			node.Attributes["unique.consul.name"])
+	} else {
+		f.logger.Printf("[WARN] fingerprint.consul: malformed Consul response prevented linking")
+	}
 
 	// If the Consul Agent was previously unavailable print a message to
 	// indicate the Agent is available now

--- a/client/fingerprint/consul_test.go
+++ b/client/fingerprint/consul_test.go
@@ -9,6 +9,7 @@ import (
 
 	"github.com/hashicorp/nomad/client/config"
 	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/stretchr/testify/assert"
 )
 
 func TestConsulFingerprint(t *testing.T) {
@@ -159,3 +160,44 @@ const mockConsulResponse = `
   }
 }
 `
+
+// TestConsulFingerprint_UnexpectedResponse asserts that the Consul
+// fingerprinter does not panic when it encounters an unexpected response.
+// See https://github.com/hashicorp/nomad/issues/3326
+func TestConsulFingerprint_UnexpectedResponse(t *testing.T) {
+	assert := assert.New(t)
+	fp := NewConsulFingerprint(testLogger())
+	node := &structs.Node{
+		Attributes: make(map[string]string),
+	}
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+		fmt.Fprintln(w, "{}")
+	}))
+	defer ts.Close()
+
+	config := config.DefaultConfig()
+	config.ConsulConfig.Addr = strings.TrimPrefix(ts.URL, "http://")
+
+	ok, err := fp.Fingerprint(config, node)
+	assert.Nil(err)
+	assert.True(ok)
+
+	attrs := []string{
+		"consul.server",
+		"consul.version",
+		"consul.revision",
+		"unique.consul.name",
+		"consul.datacenter",
+	}
+	for _, attr := range attrs {
+		if v, ok := node.Attributes[attr]; ok {
+			t.Errorf("unexpected node attribute %q with vlaue %q", attr, v)
+		}
+	}
+
+	if v, ok := node.Links["consul"]; ok {
+		t.Errorf("Unexpected link to consul: %v", v)
+	}
+}

